UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs

```diff
@@ -217,42 +229,50 @@ void SendCommandEvent(CommandProto command)
 
         public void DecideBatch()
         {
-            if (m_CurrentAgents.Values.All(l => l.Count == 0))
+            if (!m_NeedCommunicateThisStep)
             {
                 return;
             }
-            foreach (var brainKey in m_CurrentAgents.Keys)
-            {
-                using (TimerStack.Instance.Scoped("AgentInfo.ToProto"))
-                {
-                    if (m_CurrentAgents[brainKey].Count > 0)
-                    {
-                        foreach (var agent in m_CurrentAgents[brainKey])
-                        {
-                            // Update the sensor data on the AgentInfo
-                            agent.GenerateSensorData();
-                            var agentInfoProto = agent.Info.ToAgentInfoProto();
-                            m_CurrentUnityRlOutput.AgentInfos[brainKey].Value.Add(agentInfoProto);
-                        }
+            m_NeedCommunicateThisStep = false;
 
-                    }
-                }
-            }
             SendBatchedMessageHelper();
-            foreach (var brainKey in m_CurrentAgents.Keys)
-            {
-                m_CurrentAgents[brainKey].Clear();
-            }
         }
 
         /// <summary>
         /// Sends the observations of one Agent. 
         /// </summary>
         /// <param name="brainKey">Batch Key.</param>
         /// <param name="agent">Agent info.</param>
-        public void PutObservations(string brainKey, Agent agent)
+        public void PutObservations(string brainKey, AgentInfo info, List<ISensor> sensors, Action<AgentAction> action)
         {
-            m_CurrentAgents[brainKey].Add(agent);
+            if (!m_ActionCallbacks.ContainsKey(brainKey))
+            {
+                int numFloatObservations = sensors.GetSensorFloatObservationSize();
+                if (m_VectorObservationBuffer.Length < numFloatObservations)
+                {
+                    m_VectorObservationBuffer = new float[numFloatObservations];
+                }
+            }
+
+            if (!m_SensorShapeValidators.ContainsKey(brainKey))
+            {
+                m_SensorShapeValidators[brainKey] = new SensorShapeValidator();
+            }
+            m_SensorShapeValidators[brainKey].ValidateSensors(sensors);
```

Comment:
Wrap this in `#if DEBUG` (and similar in ModelRunner)? I think that's what we had around the old calls.